interfaces/builtin/polkit.go

```diff
@@ -145,11 +151,168 @@ func (iface *polkitInterface) PolkitConnectedPlug(spec *polkit.Specification, pl
 	return nil
 }
 
-func (iface *polkitInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+type polkitInstallRule struct {
+	Name, Sha3_384 string
+}
+
+func (iface *polkitInterface) parseAndValidateInstallRules(attribs interfaces.Attrer) ([]polkitInstallRule, error) {
+	var ruleEntries []map[string]string
+	if err := attribs.Attr("install-rules", &ruleEntries); err != nil {
+		return nil, err
+	}
+	if len(ruleEntries) == 0 {
+		return nil, fmt.Errorf("\"install-rules\" must have at least one entry")
+	}
+	rules := make([]polkitInstallRule, len(ruleEntries))
+	for i, ruleEntry := range ruleEntries {
+		rule := polkitInstallRule{}
+		for key, val := range ruleEntry {
+			switch key {
+			case "name":
+				if err := validate.ValidateRuleFileName(val); err != nil {
+					return nil, err
+				}
+				rule.Name = val
+			case "sha3-384":
+				rule.Sha3_384 = val
+			default:
+				return nil, fmt.Errorf("unexpected key %q for \"install-rules\" entry", key)
+			}
+		}
+		if rule.Name == "" {
+			return nil, fmt.Errorf("key \"name\" is required for \"install-rules\" entry")
+		}
+		if rule.Sha3_384 == "" {
+			return nil, fmt.Errorf("key \"sha3-384\" is required for \"install-rules\" entry")
+		}
+		rules[i] = rule
+	}
+	return rules, nil
+}
+
+func readPolkitRule(filename string, installRules []polkitInstallRule) (polkit.Rule, error) {
+	// Find matching install-rules entry
+	base := filepath.Base(filename)
+	var exptectedHash string
+	for _, installRule := range installRules {
+		if installRule.Name == base {
+			exptectedHash = installRule.Sha3_384
+			break
+		}
+	}
+	if exptectedHash == "" {
+		return nil, fmt.Errorf("no matching \"install-rule\" entry found for %q", filename)
+	}
+	// Compute sha3-384 hash of matched file
+	hashDigest, _, err := osutil.FileDigest(filename, crypto.SHA3_384)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read hash digest of %q: %v", filename, err)
```

Comment:
obtain/compute/derive
```suggestion
		return nil, fmt.Errorf("cannot obtain hash of %q: %w", filename, err)
```